### PR TITLE
extend cache interface to support xtensa coherency architecture

### PR DESCRIPTION
**Pull request by nashif**
This move custom cache calls implemented via soc APIs to a generic interface that can be used to abstract the cache calls.
Eliminate more use of z_soc_* call 

- cache: introduce coherent cache interface
- xtensa: move to use system cache API support for coherency

### CODE REVIEW

File: arch/Kconfig [hunk 1]
```diff
@@ -902,6 +907,13 @@ config ICACHE
 	help
 	  This option enables the support for the instruction cache (i-cache).
 
+config CACHE_COHERENCY
+	bool "Cache Coherence support"
+	depends on CPU_HAS_CACHE_COHERENCY
+	default y
+	help
+	  Cache coherency
```
Comment (carlocaione) [inline arch/Kconfig:915]: maybe a more verbose help?

File: include/zephyr/arch/cache.h [hunk 2]
```diff
@@ -335,6 +335,26 @@ size_t arch_icache_line_size_get(void);
...
 
+#if CONFIG_CACHE_COHERENCY  || __DOXYGEN__
+bool arch_cache_is_ptr_cached(void *ptr);
+
+#define cache_is_ptr_cached(ptr) arch_cache_is_ptr_cached(ptr)
+
+
+bool arch_cache_is_ptr_uncached(void *ptr);
+
+#define cache_is_ptr_uncached(ptr) arch_cache_is_ptr_uncached(ptr)
```
Comment (carlocaione) [inline include/zephyr/arch/cache.h:346]: remove the space between `bool` and `define` (like it is done in the following lines).

**Review by carlocaione — CHANGES_REQUESTED** [review]
nits

**Review by lgirdwood — DISMISSED** [review]
LGTM.

File: arch/Kconfig [hunk 3]
```diff
@@ -707,6 +707,11 @@ config CPU_HAS_DCACHE
 	help
 	  This hidden configuration should be selected when the CPU has a d-cache.
 
+config CPU_HAS_CACHE_COHERENCY
+	bool
+	help
+	  This hidden configuration should be selected when the CPU has cache coherency.
```
Comment (andyross) [inline arch/Kconfig:713]: Can we name this in the opposite direction: CPU_CACHE_INCOHERENT or something?  That's the notable state.  Almost everything is coherent.

File: arch/Kconfig [hunk 4]
```diff
@@ -902,6 +907,16 @@ config ICACHE
 	help
 	  This option enables the support for the instruction cache (i-cache).
 
+config CACHE_COHERENCY
```
Comment (andyross) [inline arch/Kconfig:910]: This name likewise seems not to be capturing the issue at hand, which is the double-mapping behavior where a pointer can be cheaply converted to point to the same cached/uncached memory at different locations.  Also it's sort of colliding with CONFIG_KERNEL_COHERENCE, which gates the kernel support for feature (and also has the same kind of ambiguity).

Maybe call it "CACHEMAP" or "CACHE_DOUBLEMAP" or something?

File: include/zephyr/cache.h [hunk 5]
```diff
@@ -441,6 +441,97 @@ static ALWAYS_INLINE size_t sys_cache_instr_line_size_get(void)
...
+
+	return false;
+}
+
+/**
+ * @brief Return cached pointer to a RAM address
+ *
+ * This function takes a pointer to any addressable object (either in
+ * cacheable memory or not) and returns a pointer that can be used to
+ * refer to the same memory through the L1 data cache.  Data read
```
Comment (andyross) [inline include/zephyr/cache.h:495]: Nitpick, but this documentation is a lie with the new scheme: it returns the same pointer if !CACHE_COHERENCY, which may or may not be cached depending on the hardware/OS setup, but certainly no promises are being made that access goes through an L1 data cache.

File: arch/Kconfig (hunk 3 above)
Comment (andyross) [reply arch/Kconfig:713]: Also worth gating this, or at least pointing out in documentation, that we're specifically talking about intra-CPU multiprocessing incoherence.  The setting really only makes sense, even on intel_adsp, when MP_NUM_CPUS > 1.

**Review by andyross — COMMENTED** [review]
Lots of nitpickery on the naming.  Really the use of an incoherent cache is unique to Xtensa, and the double-map trick is unique within that space to intel_adsp.  And realistically that seems unlikely to change.  Promoting this to a global API seems to not have a lot of real value, and just naming all the APIs "cache" seems likely to confuse readers who might thing it's an interface to a more typical cache flush API and not this weird trick.  Seems like it might be good to wall it off in its own API area and document the feature as a specific thing.

Also CONFIG_KERNEL_COHERENCE needs to either depend on this new gadget or be unified with it.

**Review by andyross — DISMISSED** [review]
All out of complaints, thanks

**Review by cfriedt — DISMISSED** [review]
LGTM

**Comment by marc-hb** [comment]
I would really like to submit this large (and... scary?) PR to SOF CI before merge but we're unfortunately still dealing on the previous round of regression(s) https://github.com/thesofproject/sof/issues/8818

Caching issues are obviously among the most time-consuming to investigate so it would be great to avoid a repeat of 8818 (which were apparently not tested much with SOF before merge)

There is light at the end of the tunnel thanks to what looks like a debugging breakthrough from @lyakh:
- #68366

Once we're out of that 8818 tunnel, submitting this to SOF CI will be as usual trivial.

tl;dr: how urgent is this PR?

**Comment by nashif** [comment]
> tl;dr: how urgent is this PR?

very

**Comment by nashif** [comment]
@carlocaione  please revisit, thanks.

**Comment by nashif** [comment]
> I would really like to submit this large (and... scary?)

scary how? this just a API naming change for the most part, original behavior was not changed.

**Comment by marc-hb** [comment]
> scary how? this just a API naming change for the most part, original behavior was not changed.

Thank you, I wasn't sure about this.

File: include/zephyr/arch/cache.h [hunk 6]
```diff
@@ -335,6 +335,20 @@ size_t arch_icache_line_size_get(void);
...
 
+#if CONFIG_CACHE_DOUBLEMAP  || __DOXYGEN__
+bool arch_cache_is_ptr_cached(void *ptr);
+#define cache_is_ptr_cached(ptr) arch_cache_is_ptr_cached(ptr)
+
+bool arch_cache_is_ptr_uncached(void *ptr);
+#define cache_is_ptr_uncached(ptr) arch_cache_is_ptr_uncached(ptr)
+
+void *arch_cache_cached_ptr_get(void *ptr);
+#define cache_cached_ptr(ptr) arch_cache_cached_ptr_get(ptr)
```
Comment (marc-hb) [inline include/zephyr/arch/cache.h:346]: This seems to be missing some sparse annotations, sparse fails like this:
https://github.com/thesofproject/sof/pull/8824#issuecomment-1920009003
```
include/zephyr/arch/cache.h:345:6: error: symbol 'arch_cache_cached_ptr_get' redeclared with different type (different address spaces):
include/zephyr/arch/cache.h:348:6: error: symbol 'arch_cache_uncached_ptr_get' redeclared with different type (incompatible argument 1 (different address spaces)):
include/zephyr/arch/cache.h:345:6: error: symbol 'arch_cache_cached_ptr_get' redeclared with different type (different address spaces):
include/zephyr/arch/cache.h:348:6: error: symbol 'arch_cache_uncached_ptr_get' redeclared with different type (incompatible argument 1 (different address spaces)):
include/zephyr/arch/cache.h:345:6: error: symbol 'arch_cache_cached_ptr_get' redeclared with different type (different address spaces):
include/zephyr/arch/cache.h:348:6: error: symbol 'arch_cache_uncached_ptr_get' redeclared with different type (incompatible argument 1 (different address spaces)):

```

**Review by marc-hb — DISMISSED** [review]
I cobbled together something with this, the revert in #68366 and the SOF dependencies. I don't have the HW test results yet but I already found a sparse warning, see below.
https://github.com/thesofproject/sof/pull/8824
(Because it's in a .h file the address space warning is multiplied a couple hundred times)

File: include/zephyr/arch/cache.h (hunk 6 above)
Comment (nashif) [reply include/zephyr/arch/cache.h:346]: addressed.

**Review by dcpleung — DISMISSED** [review]

File: include/zephyr/arch/cache.h (hunk 6 above)
Comment (marc-hb) [reply include/zephyr/arch/cache.h:346]: Much fewer sparse warnings now but still a good few: https://github.com/thesofproject/sof/pull/8824#pullrequestreview-1855124913

Running sparse locally is very easy thanks to @tejlmand's work:

1. Download and compile sparse with a simple `make`
2. add `sparse` to PATH
3. west build .... -DZEPHYR_SCA_VARIANT=sparse

**Review by marc-hb — COMMENTED** [review]
HW test results don't look good in https://github.com/thesofproject/sof/pull/8824 but as stated before, they don't look good already in the first place so it's practically impossible to tell whether this PR makes them better or worse or no difference.

**Review by cfriedt — DISMISSED** [review]

**Review by carlocaione — DISMISSED** [review]

**Comment by nashif** [comment]
> so it's practically impossible to tell whether this PR makes them better or worse or no difference.

This PR is not supposed to fix or deal with existing issues in SOF. This just adds a level of abstraction and removes `z_soc_` and `arch_xtensa_` usages in SOF which has the goal of running on architectures other than xtensa.

File: include/zephyr/arch/cache.h (hunk 6 above)
Comment (marc-hb) [reply include/zephyr/arch/cache.h:346]: > not getting any of those when I build SOF for example using sparse as SCA

Detailed reproduction steps:

```shell
cd zephyr
git fetch origin 3284d869565ef6a254868853e1102cdff438816e
git checkout     3284d869565ef6a254868853e1102cdff438816e
west update
west build -p -b intel_adsp_ace15_mtpm ../modules/audio/sof/app/ -DZEPHYR_SCA_VARIANT=sparse |& tee sparse.log

../modules/audio/sof/scripts/parse_sparse_output.sh < sparse.log 

...
zephyr/include/zephyr/cache.h:511:16: warning: incorrect type in return expression (different address spaces)
zephyr/include/zephyr/cache.h:511:16: warning: incorrect type in return expression (different address spaces)
zephyr/include/zephyr/cache.h:511:16: warning: incorrect type in return expression (different address spaces)
zephyr/include/zephyr/cache.h:511:16: warning: incorrect type in return expression (different address spaces)
zephyr/include/zephyr/cache.h:511:16: warning: incorrect type in return expression (different address spaces)
...
zephyr/include/zephyr/cache.h:535:16: warning: incorrect type in argument 1 (different address spaces)
zephyr/include/zephyr/cache.h:535:16: warning: incorrect type in argument 1 (different address spaces)
zephyr/include/zephyr/cache.h:535:16: warning: incorrect type in argument 1 (different address spaces)

...
```


There are still some warnings left even after `git checkout HEAD~3 && west update` (I will look at them now) but the address warnings above in `cache.h` are caused by this PR.

**Review by marc-hb — COMMENTED** [review]
> > so it's practically impossible to tell whether this PR makes them [the tests] ~better or~ worse or no difference.

> This PR is not supposed to fix or deal with existing issues in SOF. 

I meant "worse or no difference", sorry.

> This just adds a level of abstraction and removes z_soc_ and arch_xtensa_ usages in SOF which ...

I understand. Testing the SOF main branch is unfortunately not in a position to prove that and the lack of any regression yet. SOF validation status: unknown.

On the other hand, I triple checked the adress warnings from sparse in `cache.h` and they do come from this PR.

**Comment by nashif** [comment]
> On the other hand, I triple checked the adress warnings from sparse in `cache.h` and they do come from this PR.

yes, fedora seems to have an old version of sparse, installed from git now and I am getting the same thing.

File: include/zephyr/arch/cache.h (hunk 6 above)
Comment (marc-hb) [reply include/zephyr/arch/cache.h:346]: > There are still some warnings left even after git checkout HEAD~3 && west update (I will look at them now)

That's just because I forgot that we must always use `-DCONFIG_LOG_USE_VLA=n` when using sparse. I forgot because the final outcome and messages when forgetting it is `too many warnings!` which isn't very useful.

To summarize; when using `-DCONFIG_LOG_USE_VLA=n` + `modules/audio/sof/scripts/parse_sparse_output.sh`:
- before this PR: zero address space warning
- with this PR: about 1 screen long of "different address space" warnings.

**Comment by marc-hb** [comment]
> yes, fedora seems to have an old version of sparse, installed from git now and I am getting the same thing.

Mystery solved!

We use https://github.com/thesofproject/sparse which I updated recently and is not far from https://git.kernel.org/pub/scm/devel/sparse/sparse.git/log/. There's an extremely recent fix for an infinite loop.

I can update https://github.com/thesofproject/sparse if needed.

File: tests/boards/intel_adsp/smoke/src/cpus.c [hunk 7]
```diff
@@ -84,19 +84,19 @@ static void core_smoke(void *arg)
 	zassert_equal(cpu, arch_curr_cpu()->id, "wrong cpu");
 
 	/* Un/cached regions should be configured and distinct */
-	zassert_equal(&tag, arch_xtensa_cached_ptr((void *)&tag),
+	zassert_equal(&tag, sys_cache_cached_ptr_get((void *)&tag),
 		      "stack memory not cached");
-	zassert_not_equal(&tag, arch_xtensa_uncached_ptr((void *)&tag),
+	zassert_not_equal(&tag, sys_cache_uncached_ptr_get((void *)&tag),
```
Comment (lyakh) [inline tests/boards/intel_adsp/smoke/src/cpus.c:89]: argument should be `__sparse_cache`

File: tests/boards/intel_adsp/smoke/src/cpus.c [hunk 8]
```diff
@@ -84,19 +84,19 @@ static void core_smoke(void *arg)
...
 			  "stack memory not cached");
-	zassert_not_equal(&static_tag, arch_xtensa_cached_ptr((void *)&static_tag),
+	zassert_not_equal(&static_tag, sys_cache_cached_ptr_get((void *)&static_tag),
 		      "stack memory not cached");
-	zassert_equal(&static_tag, arch_xtensa_uncached_ptr((void *)&static_tag),
+	zassert_equal(&static_tag, sys_cache_uncached_ptr_get((void *)&static_tag),
```
Comment (lyakh) [inline tests/boards/intel_adsp/smoke/src/cpus.c:93]: ditto

File: tests/boards/intel_adsp/smoke/src/cpus.c [hunk 9]
```diff
@@ -84,19 +84,19 @@ static void core_smoke(void *arg)
...
 			  "stack memory not cached");
 
 	/* Un/cached regions should be working */
 	printk(" Cache behavior check\n");
-	volatile int *ctag = (volatile int *)arch_xtensa_cached_ptr((void *)&tag);
-	volatile int *utag = (volatile int *)arch_xtensa_uncached_ptr((void *)&tag);
+	volatile int *ctag = (volatile int *)sys_cache_cached_ptr_get((void *)&tag);
+	volatile int *utag = (volatile int *)sys_cache_uncached_ptr_get((void *)&tag);
```
Comment (lyakh) [inline tests/boards/intel_adsp/smoke/src/cpus.c:99]: all "uncached get" have a cached pointer as their argument, so the latter one should have `(void __sparce_cache *)` in the argument conversion

**Review by dcpleung — APPROVED** [review]

**Review by cfriedt — APPROVED** [review]
